# 30.08.23

# NUMPY
  ## Numeric-Python (Numpy), is a Python library that is used for numeric and scientific operations.
  ## It serves as a building block for many libraries available in Python.

# DATA STRUCTURES IN NUMPY

## The main DS of NumPy is the ndarray or n-dimensional array.
## It can easily deal with vector and matrix operations.

# NUMPY OBJECT CREATION
## Numpy array can be created using array() fn.
## The array() fn in Numpy returns an array object named ndarray.

## Syntax: np.array(object,dtype)
###   object - A python obj eg, a list
###   dtype - data type of obj eg, integer

In [1]:
import numpy as np

# creating 1-d array
student_marks_arr = np.array([34,64,69,54,26])
student_marks_arr

array([34, 64, 69, 54, 26])

In [ ]:
# Creating a numpy array from python list

pylist = [23,34,46,28,76]
nparr = np.array(pylist)
nparr

array([23, 34, 46, 28, 76])

In [ ]:
pylist

[23, 34, 46, 28, 76]

In [ ]:
# Shape of the array
nparr.shape

(5,)

In [ ]:
# Data type of data contained by the array
nparr.dtype

dtype('int64')

In [ ]:
list_hetero = [23,9.2,"oi"]
arr_hetero = np.array(list_hetero)
arr_hetero

array(['23', '9.2', 'oi'], dtype='<U32')

In [ ]:
arr_hetero.dtype

dtype('<U32')

In [ ]:
# Changing dtype

list_int = [4,5,7,23]
arr_float = np.array(list_int, dtype='float')
arr_float

array([ 4.,  5.,  7., 23.])

# WORKING WITH LARGE DATA

In [ ]:
# Calculating a million integers using Python lists

%%time
# used to calculate total operation time

list1=list(range(1,1000000))
list2=list(range(2,1000001))
list3=[]

for i in range(len(list1)):
  list3.append(list1[i]+list2[i])

CPU times: user 486 ms, sys: 101 ms, total: 588 ms
Wall time: 703 ms


In [ ]:
# Calculating a million integers using NumPy

%%time

import numpy as np

# creating numpy arrays of 1 million numbers
a = np.arange(1,1000000)
b = np.arange(2,1000001)
c = a+b

CPU times: user 5.18 ms, sys: 10 ms, total: 15.2 ms
Wall time: 20.1 ms


## A huge operational time gap can be observed:
  ### List: Wall time: 703 ms
  ### Numpy array: Wt: 24.4 ms